packages/grid/_modules_/grid/hooks/core/useGridStateInitialization.ts

```diff
@@ -1,33 +1,38 @@
 import React from 'react';
 import { DataGridProcessedProps } from '../../models/props/DataGridProps';
 import { GridApiRef } from '../../models/api/gridApiRef';
+import { GridApiCommon } from '../../models/api/gridApi';
 import { GridStateApi } from '../../models/api/gridStateApi';
 import { GridControlStateItem } from '../../models/controlStateItem';
 import { GridSignature } from '../utils/useGridApiEventHandler';
-import { GridState } from '../../models/gridState';
 import { GridEvents } from '../../models/events';
 import { useGridApiMethod } from '../utils';
-
-export const useGridStateInitialization = (apiRef: GridApiRef, props: DataGridProcessedProps) => {
-  const controlStateMapRef = React.useRef<Record<string, GridControlStateItem<any>>>({});
-  const [, rawForceUpdate] = React.useState<GridState>();
-
-  const updateControlState = React.useCallback<GridStateApi['unstable_updateControlState']>(
-    (controlStateItem) => {
-      const { stateId, ...others } = controlStateItem;
-
-      controlStateMapRef.current[stateId] = {
-        ...others,
-        stateId,
-      };
-    },
-    [],
-  );
-
-  const setState = React.useCallback<GridStateApi['setState']>(
+import { isFunction } from '../../utils/utils';
+
+export const useGridStateInitialization = <GridApi extends GridApiCommon>(
+  apiRef: GridApiRef<GridApi>,
+  props: DataGridProcessedProps,
+) => {
+  const controlStateMapRef = React.useRef<
+    Record<string, GridControlStateItem<GridApi['state'], any>>
+  >({});
+  const [, rawForceUpdate] = React.useState<GridApi['state']>();
+
+  const updateControlState = React.useCallback<
+    GridStateApi<GridApi['state']>['unstable_updateControlState']
+  >((controlStateItem) => {
+    const { stateId, ...others } = controlStateItem;
+
+    controlStateMapRef.current[stateId] = {
+      ...others,
+      stateId,
+    };
+  }, []);
+
+  const setState = React.useCallback<GridStateApi<GridApi['state']>['setState']>(
     (state) => {
-      let newState: GridState;
-      if (typeof state === 'function') {
+      let newState: GridApi['state'];
+      if (isFunction(state)) {
```

Comment:
We were moving away from using this helper. It doesn't bring much value compared to `typeof`.

Reply:
This PR is not ready for review
And this change is in #3648, see https://github.com/mui-org/material-ui-x/pull/3648#discussion_r792523469

Reply:
I've unresolved the discussion @flaviendelangle linked to, so that link works now.
Feel free to chime in, I'm also curious how to make `typeof` work in that case

Reply:
Clearly if I can get rid of the utility function without adding an ugly `as` in the code, I'm all in.